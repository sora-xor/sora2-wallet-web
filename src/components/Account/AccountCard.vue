<template>
  <s-card v-bind="{ shadow: 'always', size: 'small', borderRadius: 'medium', ...$attrs }" class="account-card">
    <div class="account">
      <div class="account-avatar">
        <slot name="avatar" />
      </div>
      <div class="account-details s-flex">
        <div class="account-credentials s-flex">
          <div class="account-credentials_name">
            <slot name="name" />
          </div>
          <div class="account-credentials_description">
            <slot name="description" />
          </div>
        </div>
        <slot />
      </div>
    </div>
  </s-card>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

@Component
export default class AccountCard extends Vue {}
</script>

<style lang="scss">
.account-card {
  display: flex;
  align-items: center;

  & > .el-card__body {
    flex: 1;
    max-width: 100%;
  }

  &.s-card.neumorphic {
    border-width: 1px;
    border-style: solid;
    border-color: transparent;
  }
}

.account {
  &-avatar {
    & > img {
      max-width: 100%;
    }
  }
  &-details {
    .el-button + .el-button {
      margin-left: 0;
    }
  }
}
</style>

<style scoped lang="scss">
$gap: $basic-spacing-small;
$avatar-size: 32px;

.account {
  display: flex;
  align-items: center;
  gap: $gap;

  &-avatar {
    display: flex;
    align-items: center;
    flex-shrink: 0;
    width: $avatar-size;
    height: $avatar-size;
    border-radius: 50%;
    overflow: hidden;
  }

  &-details {
    flex: 1;
    align-items: center;
    max-width: calc(100% - $gap - $avatar-size);
  }

  &-credentials {
    flex: 1;
    flex-direction: column;
    justify-content: center;
    overflow: hidden;

    &_name,
    &_description {
      overflow: hidden;
      text-overflow: ellipsis;
      letter-spacing: var(--s-letter-spacing-small);
    }

    &_name {
      font-size: var(--s-font-size-medium);
      font-weight: 600;
      line-height: var(--s-line-height-medium);
      outline: none;
      white-space: nowrap;
      text-overflow: ellipsis;
    }

    &_description {
      @include value-prefix(width, fit-content);
      @include hint-text;
      outline: none;
    }
  }
}
</style>
