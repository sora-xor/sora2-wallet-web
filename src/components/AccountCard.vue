<template>
  <s-card v-bind="{ shadow: 'always', size: 'small', borderRadius: 'medium', ...$attrs }" class="wallet-account">
    <div class="account s-flex">
      <div class="account-avatar">
        <slot name="avatar" />
      </div>
      <div class="account-details s-flex">
        <div class="account-credentials s-flex">
          <div class="account-credentials_name">
            <slot name="name" />
          </div>
          <div class="account-credentials_description">
            <slot name="description" />
          </div>
        </div>
        <slot />
      </div>
    </div>
  </s-card>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

@Component
export default class AccountCard extends Vue {}
</script>

<style lang="scss">
.account {
  &-details {
    .el-button + .el-button {
      margin-left: 0;
    }
  }
}
</style>

<style scoped lang="scss">
$avatar-margin-right: #{$basic-spacing-small};
$avatar-size: 32px;

.wallet-account.s-card {
  border: 1px solid transparent;
}

.account {
  align-items: center;

  &-avatar {
    display: flex;
    flex-shrink: 0;
    margin-right: $avatar-margin-right;
    width: $avatar-size;
    height: $avatar-size;
  }

  &-details {
    flex: 1;
    align-items: center;
    width: calc(100% - #{$avatar-size} - #{$avatar-margin-right});
  }

  &-credentials {
    flex: 1;
    flex-direction: column;
    justify-content: center;
    overflow: hidden;

    &_name,
    &_description {
      overflow: hidden;
      text-overflow: ellipsis;
      letter-spacing: var(--s-letter-spacing-small);
    }

    &_name {
      font-size: var(--s-font-size-medium);
      font-weight: 600;
      line-height: var(--s-line-height-medium);
      outline: none;
      white-space: nowrap;
      text-overflow: ellipsis;
    }

    &_description {
      @include value-prefix(width, fit-content);
      @include hint-text;
      outline: none;
    }
  }
}
</style>
