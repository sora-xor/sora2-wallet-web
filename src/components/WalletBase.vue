<template>
  <s-card primary class="base" border-radius="medium" shadow="always" size="big">
    <template #header>
      <div :class="headerClasses">
        <s-button
          v-if="showBack"
          class="base-title_back"
          type="action"
          :tooltip="t('backText')"
          @click="handleBackClick"
        >
          <s-icon name="arrows-chevron-left-rounded-24" size="28" />
        </s-button>
        <h3 class="base-title_text">
          {{ title }}
          <s-tooltip
            v-if="tooltip"
            class="base-title_tooltip"
            popper-class="info-tooltip base-title_tooltip-popper"
            border-radius="mini"
            :content="tooltip"
            placement="right"
          >
            <s-icon name="info-16" size="18px" />
          </s-tooltip>
        </h3>
        <s-button
          v-if="showAction"
          class="base-title_action"
          type="action"
          :tooltip="t(actionTooltip)"
          @click="handleActionClick"
        >
          <s-icon :name="actionIcon" size="28" />
        </s-button>
        <!-- <s-button
          v-if="showCleanHistory"
          class="base-title_trash"
          type="action"
          icon="basic-trash-24"
          :disabled="disabledCleanHistory"
          :tooltip="t('history.clearHistory')"
          @click="handleCleanHistoryClick"
        /> -->
        <s-button
          v-if="showClose"
          class="base-title_close"
          type="action"
          rounded
          icon="basic-close-24"
          :tooltip="t('closeText')"
          @click="handleCloseClick"
        >
          <s-icon name="basic-close-24" size="28" />
        </s-button>
      </div>
    </template>
    <slot />
  </s-card>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'

import TranslationMixin from './mixins/TranslationMixin'

@Component
export default class WalletBase extends Mixins(TranslationMixin) {
  @Prop({ default: '', type: String }) readonly title!: string
  @Prop({ default: '', type: String }) readonly tooltip!: string
  @Prop({ default: false, type: Boolean }) readonly showBack!: boolean
  @Prop({ default: false, type: Boolean }) readonly showAction!: boolean
  @Prop({ default: false, type: Boolean }) readonly showClose!: boolean
  @Prop({ default: '', type: String }) readonly actionTooltip!: string
  @Prop({ default: '', type: String }) readonly actionIcon!: string
  // @Prop({ default: false, type: Boolean }) readonly disabledCleanHistory!: boolean
  // @Prop({ default: false, type: Boolean }) readonly showCleanHistory!: boolean

  get headerClasses (): Array<string> {
    const cssClasses: Array<string> = ['base-title', 's-flex']
    if (this.showBack) {
      cssClasses.push('base-title--center')
    }
    // if (this.showCleanHistory) {
    //   cssClasses.push('base-title--has-history')
    // }
    // if (this.showAction && (this.showClose || this.showCleanHistory)) {
    if (this.showAction && this.showClose) {
      cssClasses.push('base-title--actions')
    }
    return cssClasses
  }

  handleBackClick (): void {
    this.$emit('back')
  }

  handleActionClick (): void {
    this.$emit('action')
  }

  handleCloseClick (): void {
    this.$emit('close')
  }

  // handleCleanHistoryClick (): void {
  //   this.$emit('cleanHistory')
  // }
}
</script>

<style lang="scss">
.base {
  .el-loading-mask {
    background-color: var(--s-color-utility-surface);
  }
}
.base-title_tooltip-popper.neumorphic.info-tooltip {
  max-width: 165px;
}
</style>

<style scoped lang="scss">
$button-size: var(--s-size-medium);

.base {
  width: $wallet-width;
  font-size: var(--s-font-size-small);
  line-height: var(--s-line-height-base);

  &-title {
    position: relative;
    height: $button-size;
    align-items: center;
    padding-right: calc(#{$button-size} + calc(var(--s-basic-spacing) * 2));
    margin-bottom: calc(var(--s-basic-spacing) * 2);
    &--center {
      padding-left: calc(#{$button-size} + calc(var(--s-basic-spacing) * 2));
      text-align: center;
    }
    &--has-history {
      .base-title_action {
        right: calc(var(--s-size-medium) + var(--s-basic-spacing));
      }
    }
    &--actions {
      &.base-title--center {
        padding-left: calc(#{$button-size} * 2 + calc(var(--s-basic-spacing) * 2));
      }
      padding-right: calc(#{$button-size} * 2 + calc(var(--s-basic-spacing) * 2));
      .base-title_action {
        right: calc(#{$button-size} + var(--s-basic-spacing));
      }
    }
    .el-button {
      position: absolute;
    }
    &_text {
      flex: 1;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      font-size: var(--s-font-size-large);
      line-height: var(--s-line-height-small);
      font-weight: 300;
      letter-spacing: var(--s-letter-spacing-mini);
    }
    &_back {
      left: 0;
    }
    &_action, &_trash, &_close {
      right: 0;
    }
    &_tooltip {
      margin-top: auto;
      margin-bottom: auto;
      margin-left: var(--s-basic-spacing);
      cursor: pointer;
    }
  }
}
</style>
