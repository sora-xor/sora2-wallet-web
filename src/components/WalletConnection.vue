<template>
  <connection-view
    :chain-api="chainApi"
    :account="account"
    :login-account="loginAccount"
    :logout-account="logoutAccount"
    :rename-account="renameAccount"
    :check-connected-account-source="checkConnectedAccountSource"
    :close-view="navigateToAccount"
  />
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator';

import { api } from '../api';
import { RouteNames } from '../consts';
import { action, getter, mutation } from '../store/decorators';

import ConnectionView from './Connection/ConnectionView.vue';
import TranslationMixin from './mixins/TranslationMixin';

import type { Route } from '../store/router/types';
import type { PolkadotJsAccount } from '../types/common';

@Component({
  components: { ConnectionView },
})
export default class WalletConnection extends Mixins(TranslationMixin) {
  @getter.account.account public account!: Nullable<PolkadotJsAccount>;

  @action.account.loginAccount public loginAccount!: (account: PolkadotJsAccount) => Promise<void>;
  @action.account.logout public logoutAccount!: (forgetAddress?: string) => Promise<void>;
  @action.account.renameAccount public renameAccount!: (data: { address: string; name: string }) => Promise<void>;
  @action.account.checkConnectedAccountSource public checkConnectedAccountSource!: (source: string) => Promise<void>;

  get chainApi() {
    return api;
  }

  @mutation.router.navigate private navigate!: (options: Route) => void;

  navigateToAccount(): void {
    this.navigate({ name: RouteNames.Wallet });
  }
}
</script>

<style lang="scss" scoped>
.learn-more-btn {
  width: 100%;
}
</style>
