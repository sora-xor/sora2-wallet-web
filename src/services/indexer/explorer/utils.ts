import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
