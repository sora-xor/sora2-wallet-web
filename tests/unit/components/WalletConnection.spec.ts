import { useDescribe, useShallowMount, useVuex } from '../../utils';
import WalletConnection from '@/components/WalletConnection.vue';
import WalletBase from '@/components/WalletBase.vue';
import AccountCard from '@/components/Account/AccountCard.vue';
import {
  POLKADOT_JS_ACCOUNTS_MOCK,
  POLKADOT_WALLET_MOCK,
  SUBWALLET_WALLET_MOCK,
} from '../../utils/WalletConnectionMock';

const createStore = ({
  currentRouteParams = { isAccountSwitch: false },
  polkadotJsAccounts = POLKADOT_JS_ACCOUNTS_MOCK,
  availableWallets = [POLKADOT_WALLET_MOCK, SUBWALLET_WALLET_MOCK],
} = {}) =>
  useVuex({
    settings: {
      state: () => ({
        isWalletLoaded: true,
      }),
    },
    account: {
      state: () => ({
        polkadotJsAccounts,
        availableWallets,
      }),
      actions: {
        importPolkadotJs: jest.fn(),
      },
    },
    router: {
      state: () => ({
        currentRouteParams,
      }),
    },
  });

useDescribe('WalletConnection.vue', WalletConnection, () => {
  it('[Extension available]: Connect account screen should be rendered', () => {
    const wrapper = useShallowMount(WalletConnection, {
      store: createStore(),
      stubs: {
        WalletBase,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('[Extension available]: Extensions should be rendered', async () => {
    const wrapper = useShallowMount(WalletConnection, {
      store: createStore(),
      stubs: {
        WalletBase,
        AccountCard,
      },
    });

    wrapper.setData({ step: 2 });

    await wrapper.vm.$nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('[Extension available]: Accounts should be rendered', async () => {
    const wrapper = useShallowMount(WalletConnection, {
      store: createStore(),
      stubs: {
        WalletBase,
      },
    });

    wrapper.setData({ step: 3 });

    await wrapper.vm.$nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('[Extension available]: No accounts, refresh screen should be rendered', async () => {
    const wrapper = useShallowMount(WalletConnection, {
      store: createStore({ polkadotJsAccounts: [] }),
      stubs: {
        WalletBase,
      },
    });

    wrapper.setData({ step: 3 });

    await wrapper.vm.$nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });
});
