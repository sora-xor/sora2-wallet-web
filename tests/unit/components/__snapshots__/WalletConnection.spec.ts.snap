// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletConnection.vue should renders correctly 1`] = `
<wallet-base-stub
  showclose="true"
  title=""
>
  <div
    class="wallet-connection"
  >
    <s-card-stub
      borderradius="medium"
      class="wallet-connection-create"
      shadow="never"
    >
      <div
        class="title"
      >
        
      </div>
       
      <div
        class="hint"
      >
        
      </div>
       
      <s-button-stub
        borderradius="small"
        class="action"
        icon=""
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="primary"
      >
        
        
      
      </s-button-stub>
    </s-card-stub>
     
    <s-card-stub
      borderradius="medium"
      class="wallet-connection-import"
      shadow="never"
    >
      <div
        class="title"
      >
        
      </div>
       
      <div
        class="hint"
      >
        
      </div>
       
      <s-button-stub
        borderradius="small"
        class="action"
        icon=""
        iconposition="left"
        nativetype="button"
        size="medium"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="primary"
      >
        
        
      
      </s-button-stub>
    </s-card-stub>
  </div>
</wallet-base-stub>
`;
